File: day11/ans.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> main;
    stack<int> min;
    
    MinStack() {    
        
    }
    
    void push(int x) {
        main.push(x);
        if((min.empty()) || (!min.empty() && x <= min.top())) {
            min.push(x);
        }
    }
    
    void pop() {
        int x = main.top();
        main.pop();
        if(min.top() == x) {
            min.pop();
        }
    }
    
    int top() {
        return main.top();
    }
    
    int getMin() {
        return min.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: day11/ans.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.actual = []
        self.min_val = []
        

    def push(self, x: int) -> None:
        self.actual.append(x)
        if not self.min_val or x < self.min_val[-1]:
            self.min_val.append(x)
        else:
            self.min_val.append(self.min_val[-1])

    def pop(self) -> None:
        self.actual.pop()
        self.min_val.pop()

    def top(self) -> int:
        return self.actual[-1]

    def getMin(self) -> int:
        return self.min_val[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
